test/functional/placeholder.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2015 The Incremental DOM Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var IncrementalDOM = require('../../index'),
+    patch = IncrementalDOM.patch,
+    elementOpen = IncrementalDOM.elementOpen,
+    elementClose = IncrementalDOM.elementClose,
+    elementVoid = IncrementalDOM.elementVoid,
+    elementPlaceholder = IncrementalDOM.elementPlaceholder;
+
+describe('placeholders', () => {
+  var container;
+
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  describe('without an existing DOM element', () => {
+    var el;
+
+    beforeEach(() => {
+      patch(container, () => {
+        el = elementPlaceholder('div', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
+      });
+    });
+
+    it('should render with the specified tag', () => {
+      expect(el.tagName).to.equal('DIV');
+    });
+
+    it('should render with static attributes', () => {
+      expect(el.className).to.equal('someClass');
+    });
+
+    it('should not render with dynamic attributes', () => {
+      expect(el.getAttribute('data-foo')).not.to.equal('bar');
+    });
+  });
+
+  describe('with an existing DOM element', () => {
+    var div;
+    var el;
+    var child;
+
+    beforeEach(() => {
+      patch(container, function() {
+        div = elementPlaceholder('div', 'placeholder');
+        patch(div, function() {
+          child = elementVoid('div', null, ['id', 'child']);
+        });
+      });
+
+      patch(container, () => {
+        el = elementPlaceholder('span', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
+      });
+    });
+
+    it('should keep the element', () => {
+      expect(el).to.equal(div);
+      expect(div.tagName).to.equal('DIV');
+    });
+
+    it('should not set static attributes', () => {
+      expect(div.getAttribute('class')).not.to.equal('someClass');
+    });
+
+    it('should not set dynamic attributes', () => {
+      expect(div.getAttribute('data-foo')).not.to.equal('bar');
```

Comment:
Actually, I believe it is the parent's concern. The outside should be controlled by the one instantiating the component with the inside controlled by the component itself.

Take for example the '<select>' element. When you declare the element, you can give it some configuration options via attributes, for example `disabled` or `multiple`. You may want to dynamically change the `disabled` attribute based on some other form control. The select attribute has some internal components. For example, it has an arrow on the right, but that is internal to the component.

There are quite a few problems with the way React does things which I would like to avoid altogether. For example, you will see people passing classes to components, which will conditionally add it to a class that they themselves define. It ends up that many components do this, but you never know which ones do it and which ones don't without going into the component itself and checking. It also ends up being extremely verbose since you sometimes need to pass a large number of properties which in turn are applied to the root element.

For Backbone, I think it is fine swapping the whole placeholder. Others may not be using Backbone, so there is no reason to artificially limit the library for that reason.